### PR TITLE
(Closes #1643) Extend Matmul2CodeTrans to support matrix-matrix

### LINKED ISSUE #1643: matrix matrix matmul to code transformation not supported
The matmul to code transformation only supports matrix vector. It needs to be extended to matrix matrix.

### PR DESCRIPTION
**Pull request by arporter**

### CODE REVIEW

**Comment by arporter** [comment]
This is ready for a first look now from either @rupertford or @sergisiso.

**Review by rupertford — CHANGES_REQUESTED** [review]
Looks good. Comments/requests are inline.

File: src/psyclone/psyir/transformations/intrinsics/matmul2code_trans.py [hunk 1]
```diff
@@ -118,6 +117,18 @@ class Matmul2CodeTrans(Operator2CodeTrans):
                 do j=1,M
                     R(i) = R(i) + A(i,j) * B(j)
 
+    For a matrix-matrix multiplication, if the dimensions of ``R``, ``A``,
+    and ``B`` are ``R(P,M)``, ``A(N,M)``, ``B(P,N)``, the MATMUL is replaced
+    with the following code:
+
+    .. code-block:: fortran
+
+        do i=1,P
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/psyir/transformations/intrinsics/matmul2code_trans.py:126]: Oh yes, switched.

File: src/psyclone/psyir/transformations/intrinsics/matmul2code_trans.py [hunk 2]
```diff
@@ -207,30 +220,30 @@ def validate(self, node, options=None):
...
                         raise NotImplementedError(
                             f"To use matmul2code_trans on matmul, only the "
-                            f"first two indices of the 1st (matrix) argument "
-                            f"are permitted to be Ranges but found "
+                            f"first two indices of the 1st argument are "
+                            f"permitted to be Ranges but found "
                             f"{type(index).__name__} at index {count+2}.")
 
-        if len(vector.symbol.shape) > 1 and not vector.children:
+        if len(vector.symbol.shape) > 2 and not vector.children:
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/psyir/transformations/intrinsics/matmul2code_trans.py:237]: Yes, better.

File: src/psyclone/psyir/transformations/intrinsics/matmul2code_trans.py [hunk 3]
```diff
@@ -99,11 +99,10 @@ def _get_array_bound(array, index):
 class Matmul2CodeTrans(Operator2CodeTrans):
     '''Provides a transformation from a PSyIR MATMUL Operator node to
     equivalent code in a PSyIR tree. Validity checks are also
-    performed. Currently only the matrix vector version of MATMUL is
-    supported.
+    performed.
 
-    If the dimensions of ``R``, ``A``, and ``B`` are ``R(N)``,
-    ``A(N,M)``, ``B(M)``, The transformation replaces:
+    For a matrix-vector multiplication, if the dimensions of ``R``, ``A``,
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/psyir/transformations/intrinsics/matmul2code_trans.py:104]: Good point, done.

File: src/psyclone/psyir/transformations/intrinsics/matmul2code_trans.py [hunk 4]
```diff
@@ -242,42 +255,64 @@ def validate(self, node, options=None):
...
+            # Check that the second dimension is a full range if it is
+            # a range.
+            if (len(vector.symbol.shape) > 1 and isinstance(vector.children[1],
+                                                            Range)
+                    and not vector.is_full_range(1)):
+                raise NotImplementedError(
+                    f"To use matmul2code_trans on matmul for a matrix-matrix "
+                    f"multiplication, the second index of the 2nd "
+                    f"argument '{vector.name}' must be a full range.")
+            if len(vector.children) > 2:
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/psyir/transformations/intrinsics/matmul2code_trans.py:269]: Oops.

File: src/psyclone/psyir/transformations/intrinsics/matmul2code_trans.py [hunk 5]
```diff
@@ -242,42 +255,64 @@ def validate(self, node, options=None):
...
-        additional dimensions (i.e. more than 2 for the matrix and
-        more than 1 for the vector) but the matrix can only have two
-        indices which are ranges and these must be the first two
-        indices and the vector can only have one index that is a range
-        and this must be the first index. Further, the ranges must be
-        for the full index space for that dimension (i.e. array
+        BinaryOperation. Each argument is permitted to have
+        additional dimensions (i.e. more than 2) but in each case it is only
+        the first two which may be ranges. Further, the ranges must currently
+        be for the full index space for that dimension (i.e. array
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/psyir/transformations/intrinsics/matmul2code_trans.py:285]: Yes, done.

File: src/psyclone/psyir/transformations/intrinsics/matmul2code_trans.py [hunk 6]
```diff
@@ -339,3 +374,79 @@ def apply(self, node, options=None):
...
+
+        # Create new i, j and ii loop iterators.
+        symbol_table = node.scope.symbol_table
+        i_loop_symbol = symbol_table.new_symbol("i", symbol_type=DataSymbol,
+                                                datatype=INTEGER_TYPE)
+        j_loop_symbol = symbol_table.new_symbol("j", symbol_type=DataSymbol,
+                                                datatype=INTEGER_TYPE)
+        ii_loop_symbol = symbol_table.new_symbol("ii", symbol_type=DataSymbol,
+                                                 datatype=INTEGER_TYPE)
+        # Create "result(i,j)"
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/psyir/transformations/intrinsics/matmul2code_trans.py:403]: Oh yes, that's nicer.

File: src/psyclone/psyir/transformations/intrinsics/matmul2code_trans.py [hunk 7]
```diff
@@ -339,3 +374,79 @@ def apply(self, node, options=None):
...
+                                   Literal("0.0", REAL_TYPE))
+        # Create j loop and add assignment and ii loop as children.
+        lower_bound, upper_bound, step = _get_array_bound(matrix1, 1)
+        jloop = Loop.create(j_loop_symbol, lower_bound, upper_bound, step,
+                            [assign, iiloop])
+        # Create i loop and add  j loop as child.
+        lower_bound, upper_bound, step = _get_array_bound(matrix2, 0)
+        iloop = Loop.create(i_loop_symbol, lower_bound, upper_bound, step,
+                            [jloop])
+        # Add the new code to the PSyIR
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/psyir/transformations/intrinsics/matmul2code_trans.py:449]: Oh yes, much better.

**Comment by arporter** [comment]
CI-permitting, this is ready for another look now.

File: src/psyclone/psyir/transformations/intrinsics/matmul2code_trans.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (rupertford) [reply src/psyclone/psyir/transformations/intrinsics/matmul2code_trans.py:150]: Thanks for switching but it also needs switching in the docstring.

**Review by rupertford — APPROVED** [review]
All looks good. I've taken the liberty of fixing a small docstring error. Approving.

**Pull request merged**